engine/sync_manager.go

```diff
@@ -397,72 +359,72 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
-			m.currencyPairs[x].Pair.Equal(p) &&
-			m.currencyPairs[x].AssetType == a {
-			switch syncType {
-			case SyncItemTicker:
-				origHadData := m.currencyPairs[x].Ticker.HaveData
-				m.currencyPairs[x].Ticker.LastUpdated = time.Now()
-				if err != nil {
-					m.currencyPairs[x].Ticker.NumErrors++
-				}
-				m.currencyPairs[x].Ticker.HaveData = true
-				m.currencyPairs[x].Ticker.IsProcessing = false
-				if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
-					removedCounter++
-					log.Debugf(log.SyncMgr, "%s ticker sync complete %v [%d/%d].",
-						exchangeName,
-						m.FormatCurrency(p).String(),
-						removedCounter,
-						createdCounter)
-					m.initSyncWG.Done()
-				}
-				return nil
-			case SyncItemOrderbook:
-				origHadData := m.currencyPairs[x].Orderbook.HaveData
-				m.currencyPairs[x].Orderbook.LastUpdated = time.Now()
-				if err != nil {
-					m.currencyPairs[x].Orderbook.NumErrors++
-				}
-				m.currencyPairs[x].Orderbook.HaveData = true
-				m.currencyPairs[x].Orderbook.IsProcessing = false
-				if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
-					removedCounter++
-					log.Debugf(log.SyncMgr, "%s orderbook sync complete %v [%d/%d].",
-						exchangeName,
-						m.FormatCurrency(p).String(),
-						removedCounter,
-						createdCounter)
-					m.initSyncWG.Done()
-				}
-				return nil
-			case SyncItemTrade:
-				origHadData := m.currencyPairs[x].Trade.HaveData
-				m.currencyPairs[x].Trade.LastUpdated = time.Now()
-				if err != nil {
-					m.currencyPairs[x].Trade.NumErrors++
-				}
-				m.currencyPairs[x].Trade.HaveData = true
-				m.currencyPairs[x].Trade.IsProcessing = false
-				if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
-					removedCounter++
-					log.Debugf(log.SyncMgr, "%s trade sync complete %v [%d/%d].",
-						exchangeName,
-						m.FormatCurrency(p).String(),
-						removedCounter,
-						createdCounter)
-					m.initSyncWG.Done()
-				}
-				return nil
-			}
+	c, exists := m.currencyPairs[k]
+	if !exists {
+		return fmt.Errorf("%w for %s %s %s", errCouldNotSyncNewData, k.Exchange, k.Pair, k.AssetType)
+	}
+
+	c.locks[syncType].Lock()
+	defer c.locks[syncType].Unlock()
+
+	if c.trackers[syncType] == nil {
+		c.trackers[syncType] = &syncBase{}
+	}
+	s := c.trackers[syncType]
+
+	if !s.IsUsingWebsocket {
+		s.IsUsingWebsocket = true
+		s.IsUsingREST = false
+		n := "Ticker"
+		if syncType == SyncItemOrderbook {
+			n = "Orderbook"
+		} else if syncType == SyncItemTrade {
+			n = "Trade"
 		}
+		log.Warnf(log.SyncMgr,
+			"%s %s %s: %s Websocket re-enabled, switching from rest to websocket",
+			c.Exchange,
+			m.FormatCurrency(c.Pair),
+			strings.ToUpper(c.AssetType.String()),
+			n,
+		)
+	}
+
+	return m.update(c, syncType, err)
+}
+
+// update notifies the syncManager to change the last updated time for a exchange asset pair
+func (m *syncManager) update(c *currencyPairSyncAgent, syncType int, err error) error {
```

Comment:
Running ./gocryptotrader has a diff in name tags between ticker and orderbook.
This branch:
```
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete BTC-USDT [1/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket BTC-USDT SPOT TICKER: Last 26429.00000000 Ask 26429.10000000 Bid 26429.00000000 High 26705.70000000 Low 26255.10000000 Volume 5387.14800189
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete WGRT-USDT [2/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket WGRT-USDT SPOT TICKER: Last 0.00306000 Ask 0.00309000 Bid 0.00306000 High 0.00318000 Low 0.00306000 Volume 132756.60727300
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete FSN-USDT [3/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket FSN-USDT SPOT TICKER: Last 0.24760000 Ask 0.24760000 Bid 0.24470000 High 0.25400000 Low 0.24290000 Volume 25332.95741200
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete LTC-USD-230623 [4/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket LTC-USD-230623 FUTURES TICKER: Last 77.42000000 Ask 77.33000000 Bid 77.26000000 High 78.44000000 Low 76.77000000 Volume 875.06350000
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete BTC-USD-230630-50000-C [5/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket BTC-USD-230630-50000-C OPTIONS TICKER: Last 0.00050000 Ask 0.00050000 Bid 0.00000000 High 0.00050000 Low 0.00050000 Volume 0.00000000
[DEBUG] | SYNC | 19/06/2023 15:04:15 | Okx ticker sync complete GRT-BTC [6/12].
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket GRT-BTC MARGIN TICKER: Last 0.00000376 Ask 0.00000375 Bid 0.00000374 High 0.00000387 Low 0.00000372 Volume 5931.00000000
[INFO]  | SYNC | 19/06/2023 15:04:15 | Okx websocket GRT-BTC MARGIN TICKER: Last 0.00000376 Ask 0.00000375 Bid 0.00000374 High 0.00000387 Low 0.00000372 Volume 5931.00000000
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete WGRT-USDT [7/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket WGRT-USDT SPOT ORDERBOOK: Bids len: 138 Amount: 133989819.001669 WGRT. Total value: 12241.253997489193 Asks len: 361 Amount: 26946275.269770 WGRT. Total value: 21553945.33721149
[WARN]  | ORDERBOOK | 19/06/2023 15:04:16 | Potential book issue for exchange Okx pair BTC-USD-230630-50000-C asset options length Bids 0 length Asks 1
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete BTC-USDT [8/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket BTC-USDT SPOT ORDERBOOK: Bids len: 400 Amount: 191.025491 BTC. Total value: 5040936.126136789 Asks len: 400 Amount: 142.042803 BTC. Total value: 3757631.5236464175
[WARN]  | ORDERBOOK | 19/06/2023 15:04:16 | Potential book issue for exchange Okx pair BTC-USD-230630-50000-C asset options length Bids 0 length Asks 1
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete BTC-USD-230630-50000-C [9/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket BTC-USD-230630-50000-C OPTIONS ORDERBOOK: Bids len: 0 Amount: 0.000000 BTC. Total value: 0 Asks len: 1 Amount: 6560.000000 BTC. Total value: 3.2800000000000002
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete LTC-USD-230623 [10/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket LTC-USD-230623 FUTURES ORDERBOOK: Bids len: 35 Amount: 45463.000000 LTC. Total value: 3488379.9800000004 Asks len: 32 Amount: 47019.000000 LTC. Total value: 3659670.2099999995
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Okx ticker sync complete GRT-BTC [11/12].
[INFO]  | SYNC | 19/06/2023 15:04:16 | Okx websocket FSN-USDT SPOT ORDERBOOK: Bids len: 107 Amount: 136425.007789 FSN. Total value: 21669.947007397797 Asks len: 400 Amount: 134407.786550 FSN. Total value: 69128.4497119405
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Exchange CurrencyPairSyncer initial sync is complete.
[DEBUG] | SYNC | 19/06/2023 15:04:16 | Exchange CurrencyPairSyncer initial sync took 1.3246107s [12 sync items].
```

Master:
```
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete GRT-BTC [1/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket GRT-BTC MARGIN TICKER: Last 0.00000376 Ask 0.00000376 Bid 0.00000374 High 0.00000387 Low 0.00000372 Volume 5931.00000000
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket GRT-BTC MARGIN TICKER: Last 0.00000376 Ask 0.00000376 Bid 0.00000374 High 0.00000387 Low 0.00000372 Volume 5931.00000000
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete BTC-USDT [2/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket BTC-USDT SPOT TICKER: Last 26428.10000000 Ask 26428.20000000 Bid 26428.10000000 High 26705.70000000 Low 26255.10000000 Volume 5384.02712319
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete WGRT-USDT [3/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket WGRT-USDT SPOT TICKER: Last 0.00306000 Ask 0.00309000 Bid 0.00306000 High 0.00318000 Low 0.00306000 Volume 132756.60727300
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete FSN-USDT [4/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket FSN-USDT SPOT TICKER: Last 0.24760000 Ask 0.24760000 Bid 0.24470000 High 0.25400000 Low 0.24290000 Volume 25332.95741200
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete BTC-USD-230630-50000-C [5/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket BTC-USD-230630-50000-C OPTIONS TICKER: Last 0.00050000 Ask 0.00050000 Bid 0.00000000 High 0.00050000 Low 0.00050000 Volume 0.00000000
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx ticker sync complete LTC-USD-230623 [6/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket LTC-USD-230623 FUTURES TICKER: Last 77.42000000 Ask 77.35000000 Bid 77.28000000 High 78.44000000 Low 76.77000000 Volume 875.06350000
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete WGRT-USDT [7/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket WGRT-USDT SPOT ORDERBOOK: Bids len: 138 Amount: 133989819.001669 WGRT. Total value: 12241.253997489193 Asks len: 361 Amount: 26946275.269770 WGRT. Total value: 21553945.33721149
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete BTC-USDT [8/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket BTC-USDT SPOT ORDERBOOK: Bids len: 400 Amount: 195.743380 BTC. Total value: 5165560.187974127 Asks len: 400 Amount: 138.014803 BTC. Total value: 3651296.436886323
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete BTC-USD-230630-50000-C [9/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket BTC-USD-230630-50000-C OPTIONS ORDERBOOK: Bids len: 0 Amount: 0.000000 BTC. Total value: 0 Asks len: 1 Amount: 6560.000000 BTC. Total value: 3.2800000000000002
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete LTC-USD-230623 [10/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket LTC-USD-230623 FUTURES ORDERBOOK: Bids len: 32 Amount: 44234.000000 LTC. Total value: 3393916.5299999993 Asks len: 31 Amount: 45708.000000 LTC. Total value: 3558045.8899999997
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete GRT-BTC [11/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket GRT-BTC MARGIN ORDERBOOK: Bids len: 27 Amount: 74502.000000 GRT. Total value: 0.23399061000000004 Asks len: 77 Amount: 218705.000000 GRT. Total value: 1.7037535199999996
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Okx orderbook sync complete FSN-USDT [12/12].
[INFO]  | SYNC | 19/06/2023 15:06:12 | Okx websocket FSN-USDT SPOT ORDERBOOK: Bids len: 107 Amount: 136425.007789 FSN. Total value: 21669.947007397797 Asks len: 400 Amount: 134407.786550 FSN. Total value: 69128.4497119405
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Exchange CurrencyPairSyncer initial sync is complete.
[DEBUG] | SYNC | 19/06/2023 15:06:12 | Exchange CurrencyPairSyncer initial sync took 1.3083655s [12 sync items].
```


```diff
diff --git a/engine/sync_manager.go b/engine/sync_manager.go
index 6bc4fe78b..d9bcc19d8 100644
--- a/engine/sync_manager.go
+++ b/engine/sync_manager.go
@@ -378,29 +378,30 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
        }
        s := c.trackers[syncType]

+       name := "Ticker"
+       if syncType == SyncItemOrderbook {
+               name = "Orderbook"
+       } else if syncType == SyncItemTrade {
+               name = "Trade"
+       }
+
        if !s.IsUsingWebsocket {
                s.IsUsingWebsocket = true
                s.IsUsingREST = false
-               n := "Ticker"
-               if syncType == SyncItemOrderbook {
-                       n = "Orderbook"
-               } else if syncType == SyncItemTrade {
-                       n = "Trade"
-               }
                log.Warnf(log.SyncMgr,
                        "%s %s %s: %s Websocket re-enabled, switching from rest to websocket",
                        c.Exchange,
                        m.FormatCurrency(c.Pair),
                        strings.ToUpper(c.AssetType.String()),
-                       n,
+                       name,
                )
        }

-       return m.update(c, syncType, err)
+       return m.update(c, syncType, name, err)
 }

 // update notifies the syncManager to change the last updated time for a exchange asset pair
-func (m *syncManager) update(c *currencyPairSyncAgent, syncType int, err error) error {
+func (m *syncManager) update(c *currencyPairSyncAgent, syncType int, name string, err error) error {
        if syncType < SyncItemTicker || syncType > SyncItemTrade {
                return fmt.Errorf("%v %w", syncType, errUnknownSyncItem)
        }
@@ -416,8 +417,9 @@ func (m *syncManager) update(c *currencyPairSyncAgent, syncType int, err error)
        s.IsProcessing = false
        if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
                removedCounter++
-               log.Debugf(log.SyncMgr, "%s ticker sync complete %v [%d/%d].",
+               log.Debugf(log.SyncMgr, "%s %s sync complete %v [%d/%d].",
                        c.Exchange,
+                       name,
                        m.FormatCurrency(c.Pair),
                        removedCounter,
                        createdCounter)
@@ -588,7 +590,7 @@ func (m *syncManager) syncTicker(c *currencyPairSyncAgent, e exchange.IBotExchan
                                relayWebsocketEvent(result, "ticker_update", c.AssetType.String(), exchangeName)
                        }
                }
-               updateErr := m.update(c, SyncItemTicker, err)
+               updateErr := m.update(c, SyncItemTicker, "Ticker", err)
                if updateErr != nil {
                        log.Errorln(log.SyncMgr, updateErr)
                }
@@ -629,7 +631,7 @@ func (m *syncManager) syncOrderbook(c *currencyPairSyncAgent, e exchange.IBotExc
                                relayWebsocketEvent(result, "orderbook_update", c.AssetType.String(), e.GetName())
                        }
                }
-               updateErr := m.update(c, SyncItemOrderbook, err)
+               updateErr := m.update(c, SyncItemOrderbook, "Orderbook", err)
                if updateErr != nil {
                        log.Errorln(log.SyncMgr, updateErr)
                }
@@ -643,7 +645,7 @@ func (m *syncManager) syncTrades(c *currencyPairSyncAgent) {
        defer c.locks[SyncItemTrade].Unlock()

        if time.Since(c.trackers[SyncItemTrade].LastUpdated) > m.config.TimeoutREST {
-               err := m.update(c, SyncItemTrade, nil)
+               err := m.update(c, SyncItemTrade, "Trade", nil)
                if err != nil {
                        log.Errorln(log.SyncMgr, err)
                }
```

Reply:
Thanks for the catch.
i don't think we want to have to pass around names like that.
Will just make syncType a type which implements Stringer instead, since we already have it.